src/inbox/mod_inbox_muc.erl

```diff
@@ -4,68 +4,92 @@
 %%% @doc
 %%%
 %%% @end
-%%% Created : 6.07.2018 
+%%% Created : 6.07.2018
 %%%-------------------------------------------------------------------
 -module(mod_inbox_muc).
 -author("dominik.stanaszek@erlang-solutions.com").
 -include("jlib.hrl").
+-include("mongoose.hrl").
 
--export([update_inbox/5, start/1, stop/1]).
+-export([update_inbox_for_muc/1, start/1, stop/1]).
 
 start(Host) ->
-    ejabberd_hooks:add(update_inbox_for_muc, Host, ?MODULE, update_inbox, 90),
-    % TODO check ooptions: if system messages stored -> 
+    ejabberd_hooks:add(update_inbox_for_muc, Host, ?MODULE, update_inbox_for_muc, 90),
+    % TODO check ooptions: if system messages stored ->
     % add hook handler for system messages on hook ie. invitation_sent
     ok.
 
 stop(Host) ->
-    ejabberd_hooks:delete(update_inbox_for_muc, Host, ?MODULE, update_inbox, 90),
+    ejabberd_hooks:delete(update_inbox_for_muc, Host, ?MODULE, update_inbox_for_muc, 90),
     ok.
 
-update_inbox(Acc, Room, {From, FromRoomJid}, AffsDict, Packet) ->
-    Affs = dict:to_list(AffsDict),
-    Users = affs_to_allowed_users(Affs),
-    FromBare = jid:to_bare(From),
-    lists:foreach(fun(User) -> update_inbox(Room, 
-                                            {FromBare, FromRoomJid}, 
-                                            User, 
-                                            Packet) end,
-                  Users),
-    Acc.
 
-affs_to_allowed_users(Users) ->
-    AllowedUsers = lists:filter(fun({_User, outcast}) -> false;
-                    (_) -> true end, Users),
-    lists:map(fun({{User, Domain, Res}, _Aff}) -> jid:make(User, Domain, Res) end,
-              AllowedUsers).
+update_inbox_for_muc(
+    #{room_jid := Room,
+      from_jid := From,
+      from_room_jid := FromRoomJid,
+      packet := Packet,
+      affiliations_map := AffsMap} = Acc) ->
+    F = fun(AffLJID, Affiliation, _) ->
+            case is_allowed_affiliation(Affiliation) of
+                true ->
+                    %% To - receiver's bare user jid (not muc jid)
+                    To = jid:make(AffLJID),
+                    %% Guess direction based on user JIDs
+                    Direction = direction(From, To),
+                    %% Host - domain of the receiver
+                    Host = To#jid.lserver,
+                    Packet2 = jlib:replace_from_to(FromRoomJid, To, Packet),
+                    update_inbox_for_user(Direction, Host, From, To, Room, Packet2);
+                false ->
+                    ok
+            end
+        end,
+    maps:fold(F, ok, AffsMap),
```

Comment:
Can we reuse `maps_foreach/2` you introduced in this PR ?

Reply:
It's more readable to use fold here, than moving maps_foreach/2 into utility module.

Reply:
ok. btw I was wondering why there is no maps:foreach/2` provided by library